### PR TITLE
feat(plugins): add command to update plugin options

### PR DESCRIPTION
**Pull request by link108**
This PR allows administrators to use halyard in order to manage plugins configurations. Configurations in halyard override the default configurations in the plugin's manifest file.

### CODE REVIEW

**Comment by link108** [comment]
@spinnakerbot assign-issue @ezimanyi

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java [hunk 1]
```diff
@@ -72,4 +90,86 @@ public InputStream downloadManifest() {
...
+      if (newMap.get(key) instanceof Map && original.get(key) instanceof Map) {
+        Map originalChild = (Map) original.get(key);
+        Map newChild = (Map) newMap.get(key);
+        original.put(key, Plugin.merge(originalChild, newChild));
+      } else if (newMap.get(key) instanceof List && original.get(key) instanceof List) {
+        List originalChild = (List) original.get(key);
+        List newChild = (List) newMap.get(key);
+        for (Object each : newChild) {
+          if (!originalChild.contains(each)) {
+            originalChild.add(each);
```
Comment (ezimanyi) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java]: This is `N^2` in the list size; not sure it matters given what I'd imagine to be typical list sizes in config files, but worth keeping in mind.

File: halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/config/model/v1/plugins/ManifestSpec.groovy [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Bol.com
+ * Copyright 2019 Armory, Inc
```
[the remark this answers is not in the document]
Comment (link108) [reply halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/config/model/v1/plugins/ManifestSpec.groovy]: This was intentional, Armory had added this file, but had incorrectly copied the copyright!

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java [hunk 3]
```diff
@@ -72,4 +90,86 @@ public InputStream downloadManifest() {
...
+  }
+
+  /**
+   * Converts YAML keys with dot notation into a nested HashMap
+   *
+   * @param key
+   * @param value
+   * @return
+   */
+  private static Map<String, Object> parseOptionHelper(String key, Object value) {
```
[the remark this answers is not in the document]
Comment (link108) [reply halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java]: Ah I should have written out a more descriptive... description, will do that for next time. The idea here is that we would pass options via the command line, eg `hal plugins edit test -Omy.nested.option=foo -Omy.nested.status=ok` and save them in the hal config. This method is required for writing the plugin overrides to the hal config in a nested way (eg to match the format of the hal config). If we do not create a nested map, then we can end up with a mix of plugin overrides in both the dot notation and nested map form, eg: 
```
plugins:
  options:
    my/plugin:
      foo.bar: baz
      foo:
        foo2: blah
        bar: not-baz
```
which is not good for both readability (eg. what is the value for foo.bar?) and does not match the rest of the hal config, as its a nested map structure. 

I was not able to find a package that creates a nested map structure. We can create flattened maps using https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/beans/factory/config/YamlProcessor.html#getFlattenedMap-java.util.Map-, which allows us to properly override the options, however we still have the issue of creating the nested map structure.  

I explored using springproperties as a way to handle these options, but again, I wasn't able to find a way to create the nested map structure from a flattened spring properties map. Using spring properties would also involve sending the options from the CLI to the daemon, which does not match the rest of Halyard.

**Comment by link108** [comment]
@ezimanyi Thank you for the comments! I've done a bit of research regarding this subject, please see https://github.com/spinnaker/halyard/pull/1412#discussion_r327859530 for details. 

I think the spirit of what we want to accomplish here still remains, spinnaker operators should be able to configure plugins, eg to override default options/configurations/properties. This PR utilizes halyard's CLI in order to add option overrides, however I think we've shown it has some issues.

I'm going to close this PR and open a new one in the near future that reads in a plugins configuration override file, that users create by hand, similar to the <service>-local.yml override files. Would appreciate your thoughts on that solution!

**Comment by ezimanyi** [comment]
@link108 : Thanks for the update!  I think that parsing a file that users have written themselves `-local.yml` sounds like a good idea.  The impression I get is that the Halyard cli commands aren't used nearly as much as people provide the config as a file, and I suspect for plugins where the command might be a bit complex to type, this might be even more the case.  (And if people move to using an operator they wouldn't even have the opportunity to use the commands.)

So all in all :+1: to taking the input as a file.  And if based on user feedback CLI parsing functionality is important, we can always add that after.

**Pull request closed without merging**